src/test/java/com/emc/object/s3/S3MetadataSearchTest.java

```diff
@@ -163,7 +163,151 @@ public void testQueryObjects() throws Exception {
         Assert.assertEquals("test", usermd.getMdMap().get("x-amz-meta-string1"));
     }
 
-    @Test // TODO: blocked by STORAGE-21341
+    @Test
+    public void testQueryObjectsWithPrefix() throws Exception {
+        String bucketName = getTestBucket();
+
+        String key1 = "prefix/object1";
+        Map<String, String> userMeta = new HashMap<String, String>();
+        userMeta.put("datetime1", "2015-01-01T00:00:00Z");
+        userMeta.put("decimal1", "3.14159");
+        userMeta.put("integer1", "42");
+        userMeta.put("string1", "test");
+
+        S3ObjectMetadata objectMetadata = new S3ObjectMetadata();
+        objectMetadata.setUserMetadata(userMeta);
+        client.putObject(new PutObjectRequest(getTestBucket(), key1, new byte[0]).withObjectMetadata(objectMetadata));
+
+        String key2 = "object2";
+
+        client.putObject(new PutObjectRequest(getTestBucket(), key2, new byte[0]).withObjectMetadata(objectMetadata));
+
+        QueryObjectsRequest request = new QueryObjectsRequest(bucketName)
+                .withAttribute("ContentType")
+                .withAttribute("Size")
+                .withQuery("(x-amz-meta-string1<='') or (x-amz-meta-string1>='')")
+                .withPrefix("prefix");
+        QueryObjectsResult result = client.queryObjects(request);
+
+        boolean is34OrLater = ecsVersion != null && ecsVersion.compareTo("3.4") >= 0;
+        boolean is37OrLater = ecsVersion != null && ecsVersion.compareTo("3.7") >= 0;
+
+        Assert.assertFalse(result.isTruncated());
+        Assert.assertEquals(bucketName, result.getBucketName());
+        if (is34OrLater)
+            Assert.assertNull(result.getNextMarker());
+        else
+            Assert.assertEquals("NO MORE PAGES", result.getNextMarker());
+        Assert.assertNotNull(result.getObjects());
+        Assert.assertEquals(1, result.getObjects().size());
+
+        QueryObject obj = result.getObjects().get(0);
+        Assert.assertEquals(key1, obj.getObjectName());
+
+        /* Blocked by STORAGE-30513 for versions before 3.7. */
+        if(is37OrLater) {
+            Assert.assertEquals(2, obj.getQueryMds().size());
+            QueryMetadata sysmd = null;
+            QueryMetadata usermd = null;
+            for (QueryMetadata m : obj.getQueryMds()) {
+                switch (m.getType()) {
+                    case SYSMD:
+                        sysmd = m;
+                        break;
+                    case USERMD:
+                        usermd = m;
+                        break;
+                }
+            }
+            System.out.println(usermd.getMdMap().toString());
```

Comment:
You don't have to print it as it's check in later codes.

Reply:
Already removed in local working copy, it is only for debug. New commit is uploaded.